tests/scanner/audit/retention_rules_engine_test.py

```diff
@@ -0,0 +1,767 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests the BigqueryRulesEngine."""
+
+import copy
+import itertools
+import json
+import mock
+import unittest
+import yaml
+from collections import namedtuple
+import tempfile
+
+from tests.unittest_utils import ForsetiTestCase
+from google.cloud.forseti.common.gcp_type import organization
+from google.cloud.forseti.common.gcp_type import project
+from google.cloud.forseti.common.util import file_loader
+from google.cloud.forseti.scanner.audit.errors import InvalidRulesSchemaError
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+from google.cloud.forseti.scanner.audit import retention_rules_engine as rre
+from google.cloud.forseti.scanner.audit import rules as scanner_rules
+from tests.unittest_utils import get_datafile_path
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+from tests.scanner.test_data import fake_retention_scanner_data as frsd
+from datetime import datetime, timedelta
+
+
+import collections
+from tests.unittest_utils import ForsetiTestCase
+from google.cloud.forseti.scanner.scanners import retention_scanner
+
+def get_rules_engine_with_rule(rule):
+    with tempfile.NamedTemporaryFile(suffix='.yaml') as f:
+        f.write(rule)
+        f.flush()
+        rules_engine = rre.RetentionRulesEngine(
+            rules_file_path=f.name)
+        rules_engine.build_rule_book()
+    return rules_engine
+
+def get_expect_violation_item(res_map, bucket_id, rule_name, rule_index):
+    RuleViolation = namedtuple(
+    'RuleViolation',
+    ['resource_name', 'resource_type', 'full_name', 'rule_name',
+     'rule_index', 'violation_type', 'violation_data', 'resource_data'])
+    lifecycle_str = json.dumps(res_map.get(bucket_id).get_lifecycle_rule())
+
+    return RuleViolation(
+        resource_name=bucket_id,
+        resource_type=res_map.get(bucket_id).type,
+        full_name=res_map.get(bucket_id).full_name,
+        rule_name=rule_name,
+        rule_index=rule_index,
+        violation_type=rre.VIOLATION_TYPE,
+        violation_data=lifecycle_str,
+        resource_data=res_map.get(bucket_id).data)
+
+
+# TODO: More tests need to be added that cover the rule attributes and how they
+#    are evaluated
+class RetentionRulesEngineTest(ForsetiTestCase):
+    """Tests for the BigqueryRulesEngine."""
+
+    def setUp(self):
+        """Set up."""
+
+    def test_invalid_rule_with_no_applies_to(self):
+        yaml_str_no_applies_to="""
+rules:
```

Comment:
YAML should handle the extra indentation fine if you add 12 spaces to keep everything looking more consistent.

(here and other constants)

Reply:
It would come out with errors if I add additional indentations. Should I write a function to delete extra indentations?